New package: booster-0.9

travankor opened the pull request:
<!-- Mark items with [x] where applicable -->

#### General
- [X] This is a new package and it conforms to the [quality requirements](https://github.com/void-linux/void-packages/blob/master/Manual.md#quality-requirements)

#### Have the results of the proposed changes been tested?
- [ ] I use the packages affected by the proposed changes on a regular basis and confirm this PR works for me
- [ ] I generally don't use the affected packages but briefly tested this PR

<!--
If GitHub CI cannot be used to validate the build result (for example, if the
build is likely to take several hours), make sure to
[skip CI](https://github.com/void-linux/void-packages/blob/master/CONTRIBUTING.md#continuous-integration).
When skipping CI, uncomment and fill out the following section.
Note: for builds that are likely to complete in less than 2 hours, it is not
acceptable to skip CI.
-->
<!-- 
#### Does it build and run successfully? 
(Please choose at least one native build and, if supported, at least one cross build. More are better.)
- [ ] I built this PR locally for my native architecture, (ARCH-LIBC)
- [ ] I built this PR locally for these architectures (if supported. mark crossbuilds):
  - [ ] aarch64-musl
  - [ ] armv7l
  - [ ] armv6l-musl
-->

ericonr commented on srcpkgs/booster/template:
```diff
@@ -0,0 +1,38 @@
+go_package="${go_import_path}/init ${go_import_path}/generator"
+go_mod_mode=readonly
+hostmakedepends="git"
+depends="busybox-static"
```
From the arch package, it would seem busybox is a dependency only for the rescue shell, and bash is a strict dependency...

Why did you go with -static, too?

travankor replied:
I'm not yet sure how bash is a dependency (Arch uses `#!/bin/bash` instead of `#!/bin/sh` for the kernel hooks but that's all I can see so far).

I would say rescue shell is core functionality, and Arch's busybox package is -static.

ericonr replied:
Fair enough. It would be nice to have a busybox virtual, if this can work with non static one.

travankor commented on srcpkgs/booster/template:
```diff
@@ -0,0 +1,38 @@
+do_build() {
+	cd generator
+	CGO_CPPFLAGS="${CPPFLAGS}" CGO_CFLAGS="${CFLAGS}" CGO_CXXFLAGS="${CXXFLAGS}" CGO_LDFLAGS="${LDFLAGS}" \
+	go build -trimpath -buildmode=pie -mod=readonly -modcacherw -ldflags "-linkmode external -extldflags"
```
Any idea on how to cross-compile this? I'm not really familiar with why these options are needed to build this.

ericonr reviewed (commented):
Is booting working ok with it? Have you had any issues?

Might make sense to add some sort of config file to the initramfs generation, like destination directory and such.

travankor commented on srcpkgs/booster/template:
```diff
@@ -0,0 +1,35 @@
+conf_files="/etc/booster.yaml"
+
+do_install() {
+	if [ "$CROSS_BUILD" ]; then
+		vbin ${GOPATH}/bin/linux_${GOARCH}/generator booster
+		vinstall ${GOPATH}/bin/linux_${GOARCH}/init 755 usr/lib/booster
+	else
+		vbin ${GOPATH}/bin/generator booster
+		vinstall ${GOPATH}/bin/init 755 usr/lib/booster
+	fi
```
The second binary needs to put in a special path not `vbin`'ed

travankor commented:
Booting works with some patches, there might be some edge cases for exotic setups. Kernel hooks handle the destination directory, and the config file is related to image generation.

UsernameRandomlyGenerated commented:
Also boots here, though I noticed it won't mount home and swap for me on LVM+LUKS, though it's probably upstream issue.

dkwo commented:
I installed it on my second machine, seems fine.

dkwo commented:
On my main machine, also using 5.12, it breaks full disk encryption: instead of prompting for password, I'm just left with an empty console.  Do you know what is going wrong?
Thanks.

travankor commented:
FDE works here, but I'm not using grub.

dkwo commented:
Works now, had to edit grub cfg, https://github.com/anatol/booster/issues/85

dkwo commented:
Ehm, while it was building fine before, now I get
```
=> booster-0.5_1: building [go] for x86_64...
   [host] git-2.31.1_1: not found
   [host] ruby-ronn-0.7.3_4: found (https://alpha.de.repo.voidlinux.org/current)
   [host] go-1.16.4_1: not found
```
and it tries to build the whole world.
Is this my issue? or does it have to do with a recent `go` update?

dkwo commented:
Other than that, it is working fine, both with encrypted and unencrypted drives, using grub.

dkwo commented:
Just a minor note: in my init, after it prompts for password, I see this, just before entering stage 1:
`runit: warning: unable to open /dev/console: file does not exist`
Do you know how to fix it?
It's not blocking, but I'd like to understand.

paper42 commented:
> Ehm, while it was building fine before, now I get
> 
> ```
> => booster-0.5_1: building [go] for x86_64...
>    [host] git-2.31.1_1: not found
>    [host] ruby-ronn-0.7.3_4: found (https://alpha.de.repo.voidlinux.org/current)
>    [host] go-1.16.4_1: not found
> ```
> 
> and it tries to build the whole world.
> Is this my issue? or does it have to do with a recent `go` update?

Rebase your branch on top of master. There have been updates to some packages since this PR was posted, xbps-src tries to build the old versions.

dkwo commented:
@paper42 Thank you.

anatol commented:
Thank you @travankor, the work looks great to me. Looking forward to see `booster` in Void Linux repo.

wrt `/dev/console` warning. I am happy to debug it, but I need a way to reproduce it locally. Ideally if I can have an up-to-date  Void Linux image (kernel + userspace) that I can start with QEMU.

dkwo commented:
Warning is now fixed for me.
I hope this will be merged soon, and perhaps even become default in Void.
Thanks.

travankor commented:
I'll add the patch for the warning and then ask for a review on IRC.

ericonr commented on srcpkgs/booster/patches/runit.patch:
```diff
@@ -0,0 +1,89 @@
+From 47041269c6a9912839e6382e32e8b37c66711ffb Mon Sep 17 00:00:00 2001
```
I'd prefer to wait for a release with this patch.

ericonr commented on srcpkgs/booster/files/kernel-hook-postrm:
```diff
@@ -0,0 +1,11 @@
+PKGNAME="$1"
+VERSION="$2"
+
+rm -f boot/initramfs-${VERSION}.img
```
@ahesford I think this is a great opportunity to start implementing your initramfs chooser stuff, in whatever format.

ericonr commented:
With kernel 4.19.192_1, I get: `open /usr/lib/modules/4.19.192_1/modules.builtin.modinfo: no such file or directory`

It built fine for 5.11.22_1 though.

I think we should document a umask policy; the final image is world-readable, so you should call `umask 0077` in the kernel install script first (see `initramfs-regenerate` hook!)

ericonr commented:
Ok, I seem to have booted into 5.11.22_1 with booster. The only mention of it in my dmesg were errors about being unable to open `/usr/lib/modules//{usbmouse.ko,hid.ko}` (the double forward slash is verbatim).

travankor commented:
I added the umask and added booster to the `initramfs-regenerate` hook. booster won't work with zfs yet, but it should work with nvidia modules.

The previous missing module issues sound like bugs for upstream to fix.

anatol commented:
For ZFS support - here is the ticket https://github.com/anatol/booster/issues/33

> in my dmesg were errors about being unable to open /usr/lib/modules//{usbmouse.ko,hid.ko}

Could you please start a ticket with repro steps and i'll debug it.

dkwo commented:
@travankor btw, is it possible to add tests to the template?

anatol commented:
Booster [has released version 0.6](https://github.com/anatol/booster/releases/tag/0.6). This release includes runit support. Please take a look and let me know if you see any issues with booster-0.6.

dkwo commented:
This works for me, thanks.

anatol commented:
Is there anything else need to be done to add booster to the packages?

The ideas behind booster is to have a simple and fast initramfs with focus on security features (such as Tang key management or TPM binding without using systemd). It also avoids uneeded dependencies like shell, sshd or systemd in initramfs to minimize security attack surface. Having booster in Void packages will be beneficial both for Void Linux users and booster project.

dkwo commented:
It's been working well for me for the past few monts, so LGTM.

paper42 commented on srcpkgs/booster/files/kernel-hook-postinst:
```diff
@@ -0,0 +1,16 @@
+fi
+
+umask 0077
+usr/bin/booster -force -output boot/initramfs-${VERSION}.img -kernelVersion ${VERSION}
+exit $?
```
```suggestion
exec usr/bin/booster -force -output boot/initramfs-${VERSION}.img -kernelVersion ${VERSION}
```

paper42 commented on srcpkgs/booster/files/kernel-hook-postinst:
```diff
@@ -0,0 +1,16 @@
+PKGNAME="$1"
+VERSION="$2"
+
+if [ ! -x usr/bin/booster ]; then
+	exit 0
+fi
```
```suggestion
[ -x usr/bin/booster ] || exit 0
```

paper42 commented on srcpkgs/booster/files/kernel-hook-postrm:
```diff
@@ -0,0 +1,11 @@
+PKGNAME="$1"
+VERSION="$2"
+
+rm -f boot/initramfs-${VERSION}.img
+exit $?
```
Would something like this work? We don't want to ignore errors and we are returning the exit code anyway?
```suggestion
exec rm boot/initramfs-${VERSION}.img
```

ahesford replied:
There's no need for `exec` and there's no need for `exit $?` anyway. The shell exits with the return code if the last command it ran.

travankor commented:
The build is failing since it needs to use go modules. This is also a blocker for go-1.17 in general since `GOPATH` is dropped.

anatol commented:
Hi, yes booster 0.7 just got released and it includes a number of new features and bugfixes. See https://github.com/anatol/booster/releases/tag/0.7

> Has anyone used booster for secure boot

Booster itself does not sign the EFI images, but there are people who was able to configure booster for secureboot with `sbupdate`. See https://github.com/anatol/booster/issues/110

dkwo commented:
Would it make sense to include in `xbps trigger` the generation of a unified kernel image, for the purpose of secure boot?
For example, we have `sbctl` that signs stuff and generates the EFI stub via `sbctl bundle` command.

dkwo commented:
I briefly tried `linux 5.15.1` today, and the boot process stops at loading the initramfs.
Has anyone succedeed in booting that?

dkwo commented:
Now it is failing to build.

> The build is failing since it needs to use go modules.

Is this what you mean?
```
# github.com/anatol/booster/init
_build-booster-xbps/src/github.com/anatol/booster/init/luks.go:65:16: undefined: luks.ClevisTokenType
```

anatol commented:
> Now it is failing to build.
> ```
> # github.com/anatol/booster/init
> _build-booster-xbps/src/github.com/anatol/booster/init/luks.go:65:16: undefined: luks.ClevisTokenType
> ```

It looks like the luks.go dependency is incorrect. You need to use module versions as specified at go.mod

anatol commented:
> I briefly tried `linux 5.15.1` today, and the boot process stops at loading the initramfs. Has anyone succedeed in booting that?

Do you use booster version 0.7?

Could you please enable debug logs and post it here or to file a ticket at booster project?

anatol commented:
I updated Void Userspace image and run booster integration tests (ArchLinux kernel + Void Linux userspace) and it works fine. So if the problem exists then it likely relates to Void kernel configuration changes.

dkwo commented:
@anatol Thank you for helping.
Let's try to make it build first, as I currently cannot build 0.7 on Void, then I will test again linux5.15.
If I add `go_mod_mode="readonly"`, then it fails with
```
# github.com/anatol/booster/generator
_build-booster-xbps/src/github.com/anatol/booster/generator/image.go:55:16: undefined: lz4.NewWriterLegacy
/usr/lib/go/pkg/tool/linux_amd64/buildid -w $WORK/b100/exe/a.out # internal
mkdir -p /builddir/booster-0.7/_build-booster-xbps/bin/
mv $WORK/b100/exe/a.out /builddir/booster-0.7/_build-booster-xbps/bin/init
rm -r $WORK/b100/
=> ERROR: booster-0.7_1: do_build: 'go get -p "$XBPS_MAKEJOBS" -x -tags "${go_build_tags}" -ldflags "${go_ldflags}" ${go_package}' exited with 2
=> ERROR:   in do_build() at common/build-style/go.sh:39
```
Did you mean something else?

anatol commented:
error `github.com/anatol/booster/generator/image.go:55:16: undefined: lz4.NewWriterLegacy` comes from the fact that pierrec/lz4 dependency that you pull does not have `lz4.NewWriterLegacy` struct. 

Upstream has it since March 2021 https://github.com/pierrec/lz4/blob/master/writer_legacy.go  And if you use go mod build system it will pick the correct dependency version for you.

I am not familiar with go build system at Void but my guess is that you do not use built-in go modules resolver and instead manage dependencies manually. In this case that dependency needs to be adjusted to match https://github.com/anatol/booster/blob/0.7/generator/go.sum.

```
github.com/pierrec/lz4 v2.6.1+incompatible h1:9UY3+iC23yxF0UfGaYrGplQ+79Rg+h/q9FV9ix19jjM=
github.com/pierrec/lz4 v2.6.1+incompatible/go.mod h1:pdkljMzZIN41W+lC3N2tnIh5sFi+IEE17M5jbnwPHcY=
```

dkwo commented:
It seems to me the build script https://github.com/void-linux/void-packages/blob/master/common/build-style/go.sh does not find the `go.mod` files.

anatol commented:
I wonder if this go.sh script looks up for `go.mod` at the project's top level i.e. at https://github.com/anatol/booster

Booster has 3 subprojects (`init/`, `generator/`, `tests/`) and keeps its `go.mod` files there..

dkwo commented:
I agree, something funny is going on with the script. @nilium Do you have any idea?

nilium commented:
@dkwo The basic Go build style assumes the common project layout with the go.mod file in the root of the repository. It sounds like this project doesn't do that, so you probably need to write custom build steps to account for the unexpected project layout.

anatol commented:
@nilium is it Go upstream style guide or Void one to have `go.mod` at the root of the project?

nilium commented:
> @nilium is it Go upstream style guide or Void one to have `go.mod` at the root of the project?

The short answer to this is that it's both. There is limited support for projects with further nested go.mod files as far as Go is concerned but there's never been a reason to support them in the build style. To the best of my knowledge, this is also the only instance I've seen of a project being organized this way, so I'm going to say that it does not fit the build style or expectations of Go in general, and that it would require special handling in the build script.

dkwo commented:
@nilium Thanks for clarifying.
@anatol @travankor I was able to build it with the following modification: can you comment whether it makes sense, and possibly update the pr?
```
diff --git a/srcpkgs/booster/template b/srcpkgs/booster/template
index 66455f8311..045d5b7864 100644
--- a/srcpkgs/booster/template
+++ b/srcpkgs/booster/template
@@ -20,12 +20,20 @@ pre_build() {
        export CGO_ENABLED=0
 }

+do_build() {
+       cd generator
+       go install -p "$XBPS_MAKEJOBS" -mod=readonly -x -tags "${go_build_tags}" -ldflags "${go_ldflags}"
+
+       cd ../init
+       go install -p "$XBPS_MAKEJOBS" -mod=readonly -x -tags "${go_build_tags}" -ldflags "${go_ldflags}"
+}
+
 do_install() {
        if [ "$CROSS_BUILD" ]; then
-               vbin ${GOPATH}/bin/linux_${GOARCH}/generator booster
+               vbin ${GOPATH}/bin/linux_${GOARCH}/booster booster
                vinstall ${GOPATH}/bin/linux_${GOARCH}/init 755 usr/lib/booster
        else
-               vbin ${GOPATH}/bin/generator booster
+               vbin ${GOPATH}/bin/booster booster
                vinstall ${GOPATH}/bin/init 755 usr/lib/booster
        fi
        ronn docs/manpage.md
```

anatol commented:
> can you comment whether it makes sense

It looks reasonable to me.

You can also move `CGO_ENABLED=0`to `init` compilation. This is an optional parameter that makes static `init` binary that in turn minimizes generated initramfs.

travankor commented:
Thanks @dkwo, I added those changes. Although I have not had a chance to test 0.7 yet.

anatol commented on srcpkgs/booster/template:
```diff
@@ -0,0 +1,48 @@
+conf_files="/etc/booster.yaml"
+
+pre_build() {
+	export CGO_ENABLED=0
```
Could you please move `CGO_ENABLED=0` to `go install` of the init binary at the line 27?

travankor replied:
`ldd /usr/lib/booster/init` not a dynamic executable

I think `init` is already static?

anatol replied:
What I meant `CGO_ENABLED=0` should be used for `init` only. It probably be harmful for `generator` binary.

travankor replied:
I see, thanks.

dkwo commented:
@travankor They're just suggesting you do
```
cd ../init
export CGO_ENABLED=0
go install -p "$XBPS_MAKEJOBS" -mod=readonly -x -tags "${go_build_tags}" -ldflags "${go_ldflags}"
```
and remove the pre_build part.

Could you try that and squash the commits?
So that we can try t merge this soon :)
I'm using 0.7 for a few days without any issue.

anatol commented:
> @travankor They're just suggesting you do
> 
> ```
> cd ../init
> export CGO_ENABLED=0
> go install -p "$XBPS_MAKEJOBS" -mod=readonly -x -tags "${go_build_tags}" -ldflags "${go_ldflags}"
> ```
> 
> and remove the pre_build part.

This.

Or even better to avoid exporting the CGO_ENABLED variable at all:

`CGO_ENABLED=0 go install -p "$XBPS_MAKEJOBS" -mod=readonly -x -tags "${go_build_tags}" -ldflags "${go_ldflags}"`

dkwo commented:
@ericonr @ahesford Do you think this can be merged?

anatol commented:
@ericonr @ahesford let me know if you have any questions wrt booster. I would like to help to answer it.

anatol commented:
fyi Alpine Linux added `booster` to its official repository and currently testing it. So far feedback has been positive.

https://pkgs.alpinelinux.org/package/edge/testing/x86/booster

dkwo commented:
After updating to 0.8, the kernel panics at boot:
```
Kernel panic - not syncing: VFS: Unable to mount root fs on unknown-block(0,0)
CPU: 0 PID: 1 Comm: swapper/0 Not tainted 5.18.9_1 #1
Hardware name: LENOVO 20F9CTO1WW/20F9CTO1WW, BIOS N1CET87W (1.55) 05/13/2022
Call Trace:
 <TASK>
 dump_stack_lvl+0x45/0x5a
 panic+0x10c/0x299
 mount_block_root+0x1c6/0x1d5
 prepare_namespace+0x136/0x165
 kernel_init_freeable+0x266/0x231
 ? rest_init+0xd0/0xd0
 kernel_init+0x16/0x20
 ret_from_fork+0x22/0x30
 </TASK>
Kernel Offset: 0x1000000 from 0xffffffff81000000 (relocation range: 0xffffffff8000000-0xffffffffbfffffff)
---[ end Kernel panic - not syncing: VFS: Unable to mount root fs on unknown-block(0,0) ]---
```
No shell available, had to press power button.
```
$ cat /etc/booster.yaml 
busybox: true
modules: i915
modules_force_load: usbkbd
```
Any ideas?

anatol commented:
The voidlinux integration tests here https://github.com/anatol/booster/blob/master/tests/voidlinux_test.go work fine with 0.8 and current booster master branch.

So I guess the issue you see is a setup/configuration one.

Could you please enable verbose mode for booster generator and kernel initrd then post both outputs here, post your dmesg output as well. See `man booster` for instructions how to enable verbose mode.

anatol commented:
Note that booster modified flags parameters. E.g. you need to use double dash with the params.

See example of the usage here https://github.com/anatol/booster/blob/master/packaging/arch/booster-install

anatol commented:
As of the flag changes see how this change affected the arch Linux package files

https://github.com/anatol/booster/commit/efa0d40e5d5e0d83cdcfd84e3e64c25994d41200

dkwo commented:
Thank you @anatol
Just changing to `booster build --force --kernel-version ${VERSION} boot/initramfs-${VERSION}.img` in a couple of places fixes the issue.

anatol commented:
cc @jacobmyers-codeninja who is another Void+booster user and hopefully can provide more feedback about this `booster` package.

oynqr commented:
I have been using booster 0.8 for about a month with a fairly standard FDE setup on x86_64-glibc, everything has been working fine.

CameronNemo commented:
@travankor Had to make these changes to get the initramfs build to work with 0.8:

```diff
diff --git a/srcpkgs/booster/files/kernel-hook-postinst b/srcpkgs/booster/files/kernel-hook-postinst
index 5c145c3150..7e6ef0a5b8 100644
--- a/srcpkgs/booster/files/kernel-hook-postinst
+++ b/srcpkgs/booster/files/kernel-hook-postinst
@@ -10,4 +10,4 @@ VERSION="$2"
 [ -x usr/bin/booster ] || exit 0
 
 umask 0077
-usr/bin/booster -force -output boot/initramfs-${VERSION}.img -kernelVersion ${VERSION}
+usr/bin/booster build --force --kernel-version ${VERSION} boot/initramfs-${VERSION}.img
diff --git a/srcpkgs/xbps-triggers/files/initramfs-regenerate b/srcpkgs/xbps-triggers/files/initramfs-regenerate
index 62c3db859e..1ba2fdf124 100755
--- a/srcpkgs/xbps-triggers/files/initramfs-regenerate
+++ b/srcpkgs/xbps-triggers/files/initramfs-regenerate
@@ -68,7 +68,7 @@ case "${INITRAMFS_GENERATOR:-dracut}" in
 			for kdir in usr/lib/modules/*; do
 				[ -d "${kdir}/kernel" ] || continue
 				kver="${kdir##*/}"
-				booster -force -output "boot/initramfs-${kver}.img" -kernelVersion "${kver}"
+				booster build --force --kernel-version "${kver}" "boot/initramfs-${kver}.img"
 			done
 		fi
 		;;
```

CameronNemo commented:
@anatol Hey, I am getting these error messages in my dmesg, but everything seems to boot alright:

```
[    0.697022] booster: finit(ghash_clmulni_intel): open /usr/lib/modules/ghash_clmulni_intel.ko: no such file or directory
[    0.697041] booster: finit(crypto_simd): open /usr/lib/modules/crypto_simd.ko: no such file or directory
```

anatol commented:
> [    0.697022] booster: finit(ghash_clmulni_intel): open /usr/lib/modules/ghash_clmulni_intel.ko: no such file or directory
[    0.697041] booster: finit(crypto_simd): open /usr/lib/modules/crypto_simd.ko: no such file or directory

It looks like a minor issue with kernel module aliasing. Could you please file a ticket at https://github.com/anatol/booster and we'll try o debug the issue.

anatol commented on srcpkgs/xbps-triggers/files/initramfs-regenerate:
```diff
@@ -61,6 +61,17 @@ case "${INITRAMFS_GENERATOR:-dracut}" in
+			for kdir in usr/lib/modules/*; do
+				[ -d "${kdir}/kernel" ] || continue
+				kver="${kdir##*/}"
+				booster -force -output "boot/initramfs-${kver}.img" -kernelVersion "${kver}"
```
This line needs to be changed to `booster build --force --kernel-version "${kver}" "boot/initramfs-${kver}.img"`

CameronNemo commented on srcpkgs/booster/files/kernel-hook-postinst:
```diff
@@ -0,0 +1,13 @@
+[ -x usr/bin/booster ] || exit 0
+
+umask 0077
+usr/bin/booster build --force --kernelVersion ${VERSION} boot/initramfs-${VERSION}.img
```
needs to be `--kernel-version`, not `--kernelVersion`

CameronNemo commented on srcpkgs/xbps-triggers/files/initramfs-regenerate:
```diff
@@ -61,6 +61,17 @@ case "${INITRAMFS_GENERATOR:-dracut}" in
+			for kdir in usr/lib/modules/*; do
+				[ -d "${kdir}/kernel" ] || continue
+				kver="${kdir##*/}"
+				booster build --force --kernelVersion "${kver}" "boot/initramfs-${kver}.img"
```
ditto

JamiKettunen commented:
Could this be bumped to `0.9`? https://github.com/anatol/booster/releases/tag/0.9

ahesford commented:
This has been in limbo a long time. One of the reasons I haven't touched it is because we didn't have a good mechanism for switching initramfs generators. Now that https://github.com/void-linux/void-packages/pull/39284 has been merged, we at least have a uniform mechanism for choosing an initramfs generator.

Please drop the xbps-triggers modification and adopt the `initramfs` alternative the way `dracut` and `mkinitcpio` do, and let's put this PR to bed.

dkwo commented:
I tried to implement this: no changes to the trigger file, and booster template becomes
```
# Template file for 'booster'
pkgname=booster
version=0.9
revision=1
build_style=go
go_import_path=github.com/anatol/booster
go_package="${go_import_path}/generator ${go_import_path}/init"
hostmakedepends="git ruby-ronn"
depends="busybox-static"
short_desc="Fast and secure initramfs generator"
maintainer="travankor <travankor@tuta.io>"
license="MIT"
homepage="https://github.com/anatol/booster"
distfiles="https://github.com/anatol/booster/archive/${version}.tar.gz"
checksum=786b89ac5f5b6fb063ff58b490525c9fda9c6f2ae53ab06416b25993bdb72ca4
conf_files="/etc/booster.yaml"
alternatives="
 initramfs:/etc/kernel.d/post-install/20-initramfs:/usr/libexec/booster/kernel-hook-postinst
 initramfs:/etc/kernel.d/post-remove/20-initramfs:/usr/libexec/booster/kernel-hook-postrm
"
do_build() {
	cd generator
	go install -p "$XBPS_MAKEJOBS" -mod=readonly -x -tags "${go_build_tags}" -ldflags "${go_ldflags}"

	cd ../init
	CGO_ENABLED=0 go install -p "$XBPS_MAKEJOBS" -mod=readonly -x -tags "${go_build_tags}" -ldflags "${go_ldflags}"
}

do_install() {
	if [ "$CROSS_BUILD" ]; then
		vbin ${GOPATH}/bin/linux_${GOARCH}/generator booster
		vinstall ${GOPATH}/bin/linux_${GOARCH}/init 755 usr/lib/booster
	else
		vbin ${GOPATH}/bin/generator booster
		vinstall ${GOPATH}/bin/init 755 usr/lib/booster
	fi
	ronn docs/manpage.md
	vman docs/manpage.1 booster.1
	echo "busybox: true" > booster.yaml
}

post_install() {
	vconf booster.yaml
	vinstall ${FILESDIR}/kernel-hook-postinst 755 usr/libexec/booster
	vinstall ${FILESDIR}/kernel-hook-postrm 755 usr/libexec/booster
	vcompletion contrib/completion/bash bash booster
	vlicense LICENSE
}
```
I noticed the following:
```
=> booster-0.9_1: running pre-pkg hook: 04-generate-runtime-deps ...
objdump: /destdir//booster-0.9/usr/bin/booster: .gnu.version_r invalid entry
objdump: warning: private headers incomplete: bad value
   SONAME: libc.so <-> musl>=1.1.24_7
```
I'll report if I have issues with next kernel updates.

ahesford commented on srcpkgs/xbps-triggers/files/initramfs-regenerate:
```diff
@@ -8,8 +8,8 @@
-# where <generator> is one of "dracut", "mkinitcpio" or "none". By default, a
-# value of "dracut" is assumed.
+# where <generator> is one of "dracut", "mkinitcpio", "booster" or "none".
+# By default, a value of "dracut" is assumed.
```
Good point, although I neglected to update the comments when rewriting the tigger so they are inaccurate anyway.

ahesford commented on srcpkgs/booster/template:
```diff
@@ -0,0 +1,49 @@
+	echo "busybox: true" > booster.yaml
+}
+
+post_install() {
```
There's not much utility in a separate `post_install` when you are already defining `do_install`, so you might as well do all of this in one place.

ahesford commented:
@travankor I pushed two commits to your branch. Please test, comment and decide whether to squash one or both.
- The first commit refactors the build and install to address DRY and eliminate unused or empty variables. (I don't know whether specifying go arguments `-ldflags ""` and `-tags ""` is the same as omitting them; the `_build_components` function might need to specify the arguments if the effect is not the same.)
- The second commit uses the default `do_build` by restoring your definition of `go_package` to refer to both components. This **seems** to work, but I haven't run the build products to confirm this.

ahesford commented:
Meh, if the choice is between "minimalism" and "avoids redundancy with our build tools", I'd kick minimalism to the curb any day of the week and twice on Sunday. How big a savings are we be talking about?

anatol commented:
Hi, the booster author here.

It is recommended to set `CGO_ENABLED=0`. There reasons for doing this are:
1. It is the configuration that is verified by the project's integration tests
2. It is the configuration used by virtually all Linux distro packages
3. As it was pointed out, it avoids including system libraries into initramfs which makes the images smaller and its loading faster.

ahesford commented:
Because upstream notes that disabling CGO, we can drop the last commit. Please test the  second commit in this PR (my first clean-up attempt) and confirm that it produces a functional initramfs.

dkwo commented:
Tested the 2nd commit, it works for me.

ahesford commented:
OK, the cleaned up template was merged in 14d62d90f784d5b42c3fab20878de56de2547f05
